Roachtest replace non rp retry

smg260 opened the pull request.

Commit:
roachtest: add api functions accepting []string over ...string

The existing function implementations are delegated to the new
ones which end in []string and not ...string. This allows us to add
functional options as the last arg in a future change.

Epic: none
Release note: none

Commit:
roachprod: replace parallelopts with runopts via cluster api

This change introduces `install.RunOptions` to be able to
configure how and whether a command is retried via `c.Run`.
The new API functions now accept a variadic `...install.RunOption`
which can be any of `WithFailSlow(..), WithRetryOpts(..),
WithRetryFn()`.

Since we have delegated the existing API functions to preserve
current behaviour, no roachtests need be updated.

Epic: none
Release note: none

Commit:
roachprod: simplify specifying retry

This commit removes the existing `install.RetryOpts` and rolls
it into the new `install.RunOptions`. `install.RetryOpts` component
properties (retry.Options, shouldRetryFn) are split into their
own options to retain familiarity of the existing retry api.

Other changes:
`WithWaitOnFail` is changed to `WithFailSlow(bool)`.
`WithRetryDisabled` convenience option added.

Epic: none
Fixes: #105753

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/111248)
<!-- Reviewable:end -->

Commit:
roachtest: remove canary.go retry logic

This change removes the existing retry logic in canary.go which is
shared by a number of roachtests and replaces it with the
roachprod retry option.

Epic: none

Release note: None

Commit:
roachtest: replace prom setup retries with default

This change removes c.RepeatRun in favour of using the
existing c.Run with the `install.WithRetryFn` option.

Epic: none

Release note: None

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=111248) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=111248) before we can accept your contribution.<br/><hr/>**Miral Gadani** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=111248) it.</sub>
